extensions/amp-app-banner/0.1/amp-app-banner.js

```diff
@@ -0,0 +1,430 @@
+
+  /** @override */
+  preconnectCallback(onLayout) {
+    this.preconnect.url('https://itunes.apple.com', onLayout);
+  }
+
+  /** @override */
+  buildCallback() {
+    // To allow layout to be scheduled.
+    setStyles(this.element, {
+      display: '',
+      visibility: 'hidden',
+    });
+
+    // We want to fallback to browser builtin mechanism when possible.
+    const viewer = viewerFor(this.win);
+    /** @private @const {boolean} */
+    this.canShowBuiltinBanner_ = !viewer.isEmbedded() && platform.isSafari();
+    if (this.canShowBuiltinBanner_) {
+      dev.info(TAG,
+          'Browser supports builtin banners. Not rendering amp-app-banner.');
+      this.hide_();
+      return;
+    }
+
+    /** @private @const {?Element} */
+    this.metaTag_ = this.win.document.head.querySelector(
+        'meta[name=apple-itunes-app]');
+    if (!this.metaTag_) {
+      this.hide_();
+      return;
+    }
+
+    user.assert(this.element.querySelector('a[open-link]'),
+        '<a open-link> is required inside %s: %s', TAG, this.element);
+
+    this.checkIfDismissed_();
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (!this.metaTag_) {
+      return Promise.resolve();
+    }
+
+    if (this.canShowBuiltinBanner_) {
+      return Promise.resolve();
+    }
+
+    this.parseIosMetaContent_(this.metaTag_.getAttribute('content'));
+    return Promise.resolve();
+  }
+
+  /**
+   * @param {string} metaContent
+   * @private
+   */
+  parseIosMetaContent_(metaContent) {
+    const parts = metaContent.replace(/\s/,'').split(',');
+    const config = {};
+    parts.forEach(part => {
+      const keyValuePair = part.split('=');
+      config[keyValuePair[0]] = keyValuePair[1];
+    });
+
+    const appId = config['app-id'];
+    const openUrl = config['app-argument'];
+    const installAppUrl = `https://itunes.apple.com/us/app/id${appId}`;
+    const openInAppUrl = openUrl || installAppUrl;
+    this.setupOpenLink_(openInAppUrl, installAppUrl);
+  }
+}
+
+
+/**
+ * @private visible for testing.
+ */
+export class AmpAndroidAppBanner extends AbstractAppBanner {
+
+  /** @override */
+  preconnectCallback(onLayout) {
+    this.preconnect.url('https://play.google.com', onLayout);
+    if (this.manifestHref_) {
+      this.preconnect.preload(this.manifestHref_);
+    }
+  }
+
+  /** @override */
+  buildCallback() {
+    // To allow layout to be scheduled.
+    setStyles(this.element, {
+      display: '',
+      visibility: 'hidden',
+    });
+
+    const viewer = viewerFor(this.win);
+    /** @private @const {?Element} */
+    this.manifestLink_ = this.win.document.head.querySelector(
+        'link[rel=manifest],link[rel=amp-manifest]');
+
+    // We want to fallback to browser builtin mechanism when possible.
+    const isChromeAndroid = platform.isAndroid() && platform.isChrome();
+    /** @private @const {boolean} */
+    this.canShowBuiltinBanner_ = !viewer.isEmbedded() && isChromeAndroid;
```

Comment:
hmmm. I don't think so. So maybe we should still run `amp-app-banner` when on CDN.